pulser/simulation/simulation.py

```diff
@@ -145,34 +154,73 @@ def prepare_dict() -> dict[str, np.ndarray]:
                     'phase': np.zeros(self._tot_duration)}
 
         def write_samples(slot: _TimeSlot,
-                          samples_dict: Mapping[str, np.ndarray]) -> None:
+                          samples_dict: Mapping[str, np.ndarray],
+                          *qid: Union[int, str]) -> None:
+            """constructs hamiltonian coefficients, taking into account, if
+                necessary, noise errors, which are local and depend
+                on the qubit's id qid"""
             _pulse = cast(Pulse, slot.type)
-            samples_dict['amp'][slot.ti:slot.tf] += _pulse.amplitude.samples
-            samples_dict['det'][slot.ti:slot.tf] += _pulse.detuning.samples
+            noise_det = 0
+            noise_amp = 1
+            # detuning offset related to bad preparation
+            # not too large, or else ODE integration errors
+            det_spam = 150
+            for noise in self._noise:
+                if noise == 'doppler':
+                    noise_det += np.random.normal(0, self._doppler_sigma)
+                # qubit qid badly prepared
+                if noise == 'SPAM' and self.spam_detune[qid[0]]:
+                    noise_det += det_spam
+                # gaussian beam for global pulses
+                if noise == 'amplitude':
+                    position = self._qdict[qid[0]]
+                    r = np.linalg.norm(position)
+                    w0 = 175.
```

Comment:
How do I do if I want to change w0 ? Or if it depends on the device ?

Reply:
It is now done through the `SimConfig` class.